File: an2/Semestrul1/Metode Avansate de Programare/Seminar3/src/repo/InMemoryRepository.java
package repo;

import models.ValidationException;
import models.Validator;

import java.util.HashMap;
import java.util.Map;

public abstract class InMemoryRepository<E extends HasID<ID>,   ID> implements Repository<E, ID> {

    private Map<ID, E> entities;
    private Validator<E> validator;

    public InMemoryRepository(Validator<E> validator) {
        entities = new HashMap<>();
        this.validator = validator;
    }

    @Override
    public E save(E entity) throws ValidationException {
        if(entity == null)
            throw new IllegalArgumentException("Entity cannot be null");
        else if(entities.containsKey(entity.getId()))
            return entities.get(entity.getId());

        validator.validate(entity);
        entities.put(entity.getId(), entity);
        return  entity;
    }

    @Override
    public E delete(ID id) {
        return null;
    }

    @Override
    public E findOne(ID id) {
        return null;
    }

    @Override
    public Iterable<E> findAll() {
        return null;
    }
}

File: an2/Semestrul1/Metode Avansate de Programare/sem11_12/repository/InMemoryRepository.cs
using Sem11_12.Model;
using Sem11_12.Model.Validator;

using System;
using System.Collections.Generic;
using System.Linq;


namespace Sem11_12.Repository
{
    class InMemoryRepository<ID, E> : IRepository<ID, E> where E : Entity<ID>
    {
        protected IValidator<E> vali;

        protected IDictionary<ID, E> entities=new Dictionary<ID, E>();

        public InMemoryRepository(IValidator<E> vali)
        {
            this.vali = vali;
        }

        public E Delete(ID id)
        {
            throw new NotImplementedException();
        }

        public IEnumerable<E> FindAll()
        {
            return entities.Values.ToList<E>();
        }

        public E FindOne(ID id)
        {
            throw new NotImplementedException();
        }

        public E Save(E entity)
        {
            if (entity == null)
                throw new ArgumentNullException("entity must not be null");
            this.vali.Validate(entity);
            if(this.entities.ContainsKey(entity.ID)){
                return entity;
            }
            this.entities[entity.ID] = entity;
            return default(E);
        }

        public E Update(E entity)
        {
            throw new NotImplementedException();
        }
    }
}
